test/distributed/checkpoint/test_hsdp_checkpoint.py

```diff
@@ -0,0 +1,212 @@
+# Owner(s): ["oncall: distributed"]
+from copy import deepcopy
+
+import torch
+import torch.distributed.checkpoint as dist_cp
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.distributed._tensor import init_device_mesh, Replicate
+
+from torch.distributed.checkpoint.default_planner import (
+    DefaultLoadPlanner,
+    DefaultSavePlanner,
+)
+
+from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
+from torch.distributed.fsdp.fully_sharded_data_parallel import (
+    ShardingStrategy,
+    StateDictType,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_utils import (
+    instantiate_parametrized_tests,
+    parametrize,
+    run_tests,
+)
+
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    with_comms,
+)
+from torch.testing._internal.distributed.checkpoint_utils import with_temp_dir
+
+
+class SimpleModel(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net1 = nn.Linear(5, 8)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(8, 4)
+        self.net3 = nn.Linear(4, 12)
+
+    def forward(self, x):
+        x = F.relu(self.net1(x))
+        x = F.relu(self.net2(x))
+        x = F.relu(self.net3(x))
+        return x
+
+    def get_input(self):
+        return torch.rand(4, 5, device="cuda")
+
+
+class SimpleModelUneven(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net1 = nn.Linear(5, 10)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(10, 15)
+        self.net3 = nn.Linear(15, 30)
+        self.net4 = nn.Linear(30, 5)
+
+    def forward(self, x):
+        x = F.relu(self.net1(x))
+        x = F.relu(self.net2(x))
+        x = F.relu(self.net3(x))
+        x = F.relu(self.net4(x))
+        return x
+
+    def get_input(self):
+        return torch.rand(4, 5, device="cuda")
+
+
+class TestHSDPCheckpoint(DTensorTestBase):
+    @property
+    def backend(self):
+        return "cpu:gloo,cuda:nccl"
+
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    @with_temp_dir
+    @parametrize("is_even_sharded_model", [True, False])
+    def test_hsdp_checkpoint(self, is_even_sharded_model) -> None:
+        CHECKPOINT_DIR = self.temp_dir
+        simple_model = SimpleModel if is_even_sharded_model else SimpleModelUneven
+
+        mesh_2d = init_device_mesh(self.device_type, (2, self.world_size // 2))
```

Comment:
@ eracah, @mvpatel2000 

`init_device_mesh` is a new experimental feature to initialize your global mesh topology. 

When you initialize FSDP with device_mesh, you would get state_dict with DTensor instead of ShardedTensor. 

If you are using FSDP without device_mesh, you would get state_dict with ShardedTensor. ShardedTensor is not aware of the global layout so the Shard group of FSDP does not know it is replicated across nodes. So you cannot do a cluster wise-saving use DCP, as you would be passing only the pg for the first shard_group (see arg below). 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/state_dict_saver.py#L22 If you are using stable build, you would probably need to use this for the time being. 

If you want to try out the new DeviceMesh approach, you can refer to this unit test.

Reply:
Distributed Checkpoint